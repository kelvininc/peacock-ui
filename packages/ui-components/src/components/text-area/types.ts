import { EventEmitter } from '@stencil/core';
import { EIconName, EOtherIconName } from '../icon/icon.types';

export interface ITextArea {
	/** (optional) Icon to show to the left of the text field */
	icon?: EIconName | EOtherIconName;
	/** (optional) The text to show inside the text area */
	text?: string;
	/** (optional) The placeholder to show in the text area */
	placeholder?: string;
	/** (required) The maximum number of characters allowed */
	maxCharLength: number;
}

export interface ITextAreaEvents {
	/** Emits the current text when there's a change */
	textChange: EventEmitter<string>;
	/** Emits the current text when the text area loses focus */
	textChangeBlur: EventEmitter<string>;
}
