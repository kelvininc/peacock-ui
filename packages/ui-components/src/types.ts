export * from './components';
export * from './utils/types';
export * from './components/action-button/action-button.types';
export * from './components/breadcrumb-item/breadcrumb-item.types';
export * from './components/illustration/illustration.types';
export * from './components/tab-navigation/tab-navigation.types';
export * from './components/text-field/text-field.types';
export * from './components/tooltip/tooltip.types';
export * from './components/summary-card/summary-card.types';
export * from './components/search/search.types';
