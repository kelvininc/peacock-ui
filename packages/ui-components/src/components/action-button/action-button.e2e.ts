import { newE2EPage } from '@stencil/core/testing';

xdescribe('action-button', () => {
	it('renders', async () => {
		const page = await newE2EPage();

		await page.setContent('<action-button></action-button>');
		const element = await page.find('action-button');
		expect(element).toHaveClass('hydrated');
	});
});
