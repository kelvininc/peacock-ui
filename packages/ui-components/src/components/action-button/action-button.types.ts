export enum EActionButtonType {
	PrimaryButton = 'primary-button',
	PrimaryDangerButton = 'primary-danger-button',
	SecondaryButton = 'secondary-button',
	TertiaryButton = 'tertiary-button'
}
