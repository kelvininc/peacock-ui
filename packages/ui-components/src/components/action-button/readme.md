# *<kv-action-button>*



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Primary -->
<kv-action-button [type]="EActionButtonType.Primary">Primary Button</kv-action-button>

<!-- Secondary -->
<kv-action-button [type]="EActionButtonType.Secondary">Secondary Button</kv-action-button>

<!-- Tertiary -->
<kv-action-button [type]="EActionButtonType.Tertiary">Tertiary Button</kv-action-button>

<!-- Disabled -->
<kv-action-button disabled [type]="EActionButtonType.Primary">Disabled Button</kv-action-button>

<!-- Anchor -->
<kv-action-button href="/link-to-url" target="_blank" [type]="EActionButtonType.Primary">Link Button</kv-action-button>
```


### React

```tsx
import React from 'react';

import { KvActionButton } from '@kelvininc/react-ui-components';

export const SwitchButtonExample: React.FC = () => (
	<>
		{/*-- Primary --*/}
		<KvActionButton type={EActionButtonType.Primary}>Primary Button</KvActionButton>

		{/*--Secondary --*/}
		<KvActionButton type={EActionButtonType.Secondary}>Secondary Button</KvActionButton>

		{/*-- Tertiary --*/}
		<KvActionButton type={EActionButtonType.Tertiary}>Tertiary Button</KvActionButton>

		{/*-- Tertiary --*/}
		<KvActionButton type={EActionButtonType.Ghost}>Ghost Button</KvActionButton>

		{/*-- Disabled --*/}
		<KvActionButton disabled type={EActionButtonType.Primary}>
			Disabled Button
		</KvActionButton>

		{/*-- Anchor --*/}
		<KvActionButton href="/link-to-url" target="_blank" type={EActionButtonType.Primary}>
			Anchor Button
		</KvActionButton>
	</>
);
```



## Properties

| Property            | Attribute  | Description                                                                                                     | Type                                                                                                                                            | Default                |
| ------------------- | ---------- | --------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------- |
| `active`            | `active`   | (optional) If `true` the button is active                                                                       | `boolean`                                                                                                                                       | `false`                |
| `disabled`          | `disabled` | (optional) If `true` the button is disabled                                                                     | `boolean`                                                                                                                                       | `false`                |
| `download`          | `download` | (optional) Specifies that the target will be downloaded when a user clicks on. The value should be the filename | `string`                                                                                                                                        | `undefined`            |
| `href`              | `href`     | (optional) The anchor's link to open when clicking                                                              | `string`                                                                                                                                        | `undefined`            |
| `loading`           | `loading`  | (optional) If `true` the button is of type loading                                                              | `boolean`                                                                                                                                       | `false`                |
| `size`              | `size`     | (optional) Button's size                                                                                        | `EComponentSize.Large \| EComponentSize.Small`                                                                                                  | `EComponentSize.Large` |
| `target`            | `target`   | (optional) The anchor's target                                                                                  | `EAnchorTarget.BrowserDefault \| EAnchorTarget.NewTab \| EAnchorTarget.Parent \| EAnchorTarget.Top`                                             | `undefined`            |
| `type` _(required)_ | `type`     | (required) Button's type                                                                                        | `EActionButtonType.Danger \| EActionButtonType.Ghost \| EActionButtonType.Primary \| EActionButtonType.Secondary \| EActionButtonType.Tertiary` | `undefined`            |


## Events

| Event         | Description                           | Type                      |
| ------------- | ------------------------------------- | ------------------------- |
| `blurButton`  | Emitted when action button is blur    | `CustomEvent<FocusEvent>` |
| `clickButton` | Emitted when action button is clicked | `CustomEvent<MouseEvent>` |
| `focusButton` | Emitted when action button is focused | `CustomEvent<FocusEvent>` |


## Shadow Parts

| Part       | Description        |
| ---------- | ------------------ |
| `"button"` | The action button. |


## CSS Custom Properties

| Name                                    | Description                                                     |
| --------------------------------------- | --------------------------------------------------------------- |
| `--background-color-active-danger`      | Background color when state is active and theme is danger.      |
| `--background-color-active-primary`     | Background color when state is active and theme is primary.     |
| `--background-color-active-secondary`   | Background color when state is active and theme is secondary.   |
| `--background-color-active-tertiary`    | Background color when state is active and theme is tertiary.    |
| `--background-color-default-danger`     | Background color when state is default and theme is danger.     |
| `--background-color-default-primary`    | Background color when state is default and theme is primary.    |
| `--background-color-default-secondary`  | Background color when state is default and theme is secondary.  |
| `--background-color-default-tertiary`   | Background color when state is default and theme is tertiary.   |
| `--background-color-disabled-danger`    | Background color when state is disabled and theme is danger.    |
| `--background-color-disabled-primary`   | Background color when state is disabled and theme is primary.   |
| `--background-color-disabled-secondary` | Background color when state is disabled and theme is secondary. |
| `--background-color-disabled-tertiary`  | Background color when state is disabled and theme is tertiary.  |
| `--background-color-focus-danger`       | Background color when state is focus and theme is danger.       |
| `--background-color-focus-primary`      | Background color when state is focus and theme is primary.      |
| `--background-color-focus-secondary`    | Background color when state is focus and theme is secondary.    |
| `--background-color-focus-tertiary`     | Background color when state is focus and theme is tertiary.     |
| `--background-color-hover-danger`       | Background color when state is hover and theme is danger.       |
| `--background-color-hover-primary`      | Background color when state is hover and theme is primary.      |
| `--background-color-hover-secondary`    | Background color when state is hover and theme is secondary.    |
| `--background-color-hover-tertiary`     | Background color when state is hover and theme is tertiary.     |
| `--border-color-active-danger`          | Border color when state is active and theme is danger.          |
| `--border-color-active-primary`         | Border color when state is active and theme is primary.         |
| `--border-color-active-secondary`       | Border color when state is active and theme is secondary.       |
| `--border-color-active-tertiary`        | Border color when state is active and theme is tertiary.        |
| `--border-color-default-danger`         | Border color when state is default and theme is danger.         |
| `--border-color-default-primary`        | Border color when state is default and theme is primary.        |
| `--border-color-default-secondary`      | Border color when state is default and theme is secondary.      |
| `--border-color-default-tertiary`       | Border color when state is default and theme is tertiary.       |
| `--border-color-disabled-danger`        | Border color when state is disabled and theme is danger.        |
| `--border-color-disabled-primary`       | Border color when state is disabled and theme is primary.       |
| `--border-color-disabled-secondary`     | Border color when state is disabled and theme is secondary.     |
| `--border-color-disabled-tertiary`      | Border color when state is disabled and theme is tertiary.      |
| `--border-color-focus-danger`           | Border color when state is focus and theme is danger.           |
| `--border-color-focus-primary`          | Border color when state is focus and theme is primary.          |
| `--border-color-focus-secondary`        | Border color when state is focus and theme is secondary.        |
| `--border-color-focus-tertiary`         | Border color when state is focus and theme is tertiary.         |
| `--border-color-hover-danger`           | Border color when state is hover and theme is danger.           |
| `--border-color-hover-primary`          | Border color when state is hover and theme is primary.          |
| `--border-color-hover-secondary`        | Border color when state is hover and theme is secondary.        |
| `--border-color-hover-tertiary`         | Border color when state is hover and theme is tertiary.         |
| `--button-height-large`                 | Button's height when size large.                                |
| `--button-height-small`                 | Button's height when size small.                                |
| `--button-padding-x-large`              | Button's horizontal padding when size is large.                 |
| `--button-padding-x-small`              | Button's horizontal padding when size is small.                 |
| `--text-color-active-danger`            | Text color when state is active and theme is danger.            |
| `--text-color-active-primary`           | Text color when state is active and theme is primary.           |
| `--text-color-active-secondary`         | Text color when state is active and theme is secondary.         |
| `--text-color-active-tertiary`          | Text color when state is active and theme is tertiary.          |
| `--text-color-default-danger`           | Text color when state is default and theme is danger.           |
| `--text-color-default-primary`          | Text color when state is default and theme is primary.          |
| `--text-color-default-secondary`        | Text color when state is default and theme is secondary.        |
| `--text-color-default-tertiary`         | Text color when state is default and theme is tertiary.         |
| `--text-color-disabled-danger`          | Text color when state is disabled and theme is danger.          |
| `--text-color-disabled-primary`         | Text color when state is disabled and theme is primary.         |
| `--text-color-disabled-secondary`       | Text color when state is disabled and theme is secondary.       |
| `--text-color-disabled-tertiary`        | Text color when state is disabled and theme is tertiary.        |
| `--text-color-focus-danger`             | Text color when state is focus and theme is danger.             |
| `--text-color-focus-primary`            | Text color when state is focus and theme is primary.            |
| `--text-color-focus-secondary`          | Text color when state is focus and theme is secondary.          |
| `--text-color-focus-tertiary`           | Text color when state is focus and theme is tertiary.           |
| `--text-color-hover-danger`             | Text color when state is hover and theme is danger.             |
| `--text-color-hover-primary`            | Text color when state is hover and theme is primary.            |
| `--text-color-hover-secondary`          | Text color when state is hover and theme is secondary.          |
| `--text-color-hover-tertiary`           | Text color when state is hover and theme is tertiary.           |


## Dependencies

### Used by

 - [kv-action-button-icon](../action-button-icon)
 - [kv-action-button-split](../action-button-split)
 - [kv-action-button-text](../action-button-text)

### Graph
```mermaid
graph TD;
  kv-action-button-icon --> kv-action-button
  kv-action-button-split --> kv-action-button
  kv-action-button-text --> kv-action-button
  style kv-action-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


