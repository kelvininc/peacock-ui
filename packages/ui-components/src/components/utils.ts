import { isNil } from 'lodash-es';
import { DEFAULT_CONFIG } from '../globals/config';
import { UIComponentsConfig } from '../types';

export const getConfig = (): UIComponentsConfig => {
	if (typeof window !== 'undefined') {
		const KvUiComponents = window.KvUiComponents;
		if (!isNil(KvUiComponents) && !isNil(KvUiComponents.config)) {
			return KvUiComponents.config;
		}
	}

	return DEFAULT_CONFIG;
};
