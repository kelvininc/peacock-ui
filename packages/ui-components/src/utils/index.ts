export * from './types';
export * as arraysHelper from './arrays.helper';
export * as clipboardHelper from './clipboard.helper';
export * as cssClassHelper from './css-class.helper';
export * as dateHelper from './date.helper';
export * as searchHelper from './search.helper';
export * as stringHelper from './string.helper';
export * as floatingUIHelper from './floating-ui.helper';
export * as mouseEventHelper from './mouse-event.helper';
