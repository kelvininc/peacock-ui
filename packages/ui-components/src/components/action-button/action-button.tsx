import { Component, Event, EventEmitter, h, Host, Prop } from '@stencil/core';
import { EActionButtonType, IButton, IButtonEvents } from './action-button.types';
import { EAnchorTarget, EComponentSize, IAnchor } from '../../utils/types';

/**
 * @part button - The action button.
 */
@Component({
	tag: 'kv-action-button',
	styleUrl: 'action-button.scss',
	shadow: true
})
export class KvActionButton implements IButton, IButtonEvents, IAnchor {
	/** @inheritdoc */
	@Prop({ reflect: true }) type?: EActionButtonType;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) active: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) loading: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) size: EComponentSize = EComponentSize.Large;
	/** @inheritdoc */
	@Prop({ reflect: true }) href?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) target?: EAnchorTarget;
	/** @inheritdoc */
	@Prop({ reflect: true }) download?: string;

	/** @inheritdoc */
	@Event() clickButton: EventEmitter<MouseEvent>;
	/** @inheritdoc */
	@Event() focusButton: EventEmitter<FocusEvent>;
	/** @inheritdoc */
	@Event() blurButton: EventEmitter<FocusEvent>;

	private onClickButton = (event: MouseEvent) => {
		if (this.disabled) {
			return;
		}

		this.clickButton.emit(event);
	};
	private onFocusButton = (event: FocusEvent) => {
		this.focusButton.emit(event);
	};
	private onBlurButton = (event: FocusEvent) => {
		this.blurButton.emit(event);
	};

	render() {
		return (
			<Host aria-disabled={this.disabled} onClick={this.onClickButton}>
				<a
					class={{
						'action-button': true,
						'action-button--disabled': this.disabled,
						'action-button--active': this.active,
						'action-button--loading': this.loading,
						[`action-button--type-${this.type}`]: true,
						[`action-button--size-${this.size}`]: true
					}}
					tabIndex={this.disabled ? -1 : 0}
					part="button"
					download={this.download}
					href={this.href}
					target={this.target}
					onFocus={this.onFocusButton}
					onBlur={this.onBlurButton}
				>
					<slot />
				</a>
			</Host>
		);
	}
}
