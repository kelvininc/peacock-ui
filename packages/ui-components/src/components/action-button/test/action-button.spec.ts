import { SpecPage } from '@stencil/core/internal';
import { KvActionButton } from '../action-button';
import { newSpecPage } from '@stencil/core/testing';
import { EComponentSize } from '../../../utils/types';

describe('Action Button (unit tests)', () => {
	let page: SpecPage;
	let component: KvActionButton;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvActionButton],
				html: '<kv-action-button type="primary"></kv-action-button>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});

		it('should initialize `disabled` with false', () => {
			expect(component.disabled).toBe(false);
		});

		it('should initialize `active` with false', () => {
			expect(component.active).toBe(false);
		});

		it('should initialize `size` with large', () => {
			expect(component.size).toBe(EComponentSize.Large);
		});
	});
});
