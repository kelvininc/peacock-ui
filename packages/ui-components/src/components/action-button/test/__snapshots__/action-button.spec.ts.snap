// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Action Button (unit tests) when uses default props should match the snapshot 1`] = `
<kv-action-button size="large" type="primary">
  <mock:shadow-root>
    <div class="action-button-wrapper">
      <a class="action-button action-button--size-large action-button--type-primary" part="button" tabindex="0">
        <slot></slot>
      </a>
    </div>
  </mock:shadow-root>
</kv-action-button>
`;
